### PR TITLE
bpo-43478: Restrict use of Mock objects as specs

### PR DESCRIPTION
**Pull request by msuozzo**
https://bugs.python.org/issue43478

### CODE REVIEW

**Comment by msuozzo** [comment]
@gpshead

**Comment by gpshead** [comment]
When doing our internal codebase cleanups for this, how many issues did you uncover in third party open source projects?  Were they turned into PRs?  If so, noting that in the bug with links to some would be good.

Otherwise if there were a meaningful number of those, I wonder if we shouldn't make this issue a warning for 3.11 and turn it into this error for 3.12. (The thing we're aiming to prevent is an inability to test major Python packages on 3.11 alphas and betas.)  Going ahead with this as is probably makes sense.  But if it blocks testing while upstream projects fix things, we should probably turn it into a warning for one release while Issues/PRs elsewhere get filed.

**Comment by gpshead** [comment]
Can you also add an entry to the 3.11 What's New doc?  More people read that than the detailed NEWS.

**Comment by msuozzo** [comment]
> When doing our internal codebase cleanups for this, how many issues did you uncover in third party open source projects? Were they turned into PRs? If so, noting that in the bug with links to some would be good.

This last cleanup only had one true python hit: https://github.com/forseti-security/forseti-security/pull/3875

The previous PR hit tensorflow, google-research, [ocp-diag-core](https://github.com/opencomputeproject/ocp-diag-core) and a few other third-party repos but I don't have the associated PRs handy.

> Otherwise if there were a meaningful number of those, I wonder if we shouldn't make this issue a warning for 3.11 and turn it into this error for 3.12. (The thing we're aiming to prevent is an inability to test major Python packages on 3.11 alphas and betas.) Going ahead with this as is probably makes sense. But if it blocks testing while upstream projects fix things, we should probably turn it into a warning for one release while Issues/PRs elsewhere get filed.

The more expansive version of this check shipped in 3.10 with this edge case not in it. I think this is a lot less likely that this addition trips people up and breaks tests so I'm inclined to leave it in as-is.

> Can you also add an entry to the 3.11 What's New doc? More people read that than the detailed NEWS.

I will if you think it's worth it. Again, I think the part of this that shipped in 3.10 was more likely to break people.

**Comment by gpshead** [comment]
Alright, looking things over given the history I don't think we needs a What's New entry just for this.  It seems pretty low surprise and doesn't appear to have a wide blast radius of lurking problem exposure.

File: Lib/unittest/test/testmock/testmock.py [hunk 1]
```diff
@@ -226,6 +226,14 @@ class B(object):
             with self.assertRaisesRegex(InvalidSpecError,
                                         "Cannot spec attr 'B' as the spec_set "):
                 mock.patch.object(A, 'B', spec_set=A.B).start()
+            with self.assertRaisesRegex(InvalidSpecError,
+                                        "Cannot spec attr 'B' as the spec_set "):
+                mock.patch.object(A, 'B', spec_set=A.B).start()
```
Comment (gpshead) [inline Lib/unittest/test/testmock/testmock.py:231]: this stanza looks identical to the test immediately above it?  mis-paste or patch issue?

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Lib/unittest/test/testmock/testmock.py (hunk 1 above)
Comment (msuozzo) [reply Lib/unittest/test/testmock/testmock.py:231]: yep messed up patch. fixed.

**Review by gpshead — APPROVED** [review]

**Review by cjw296 — APPROVED** [review]

**Comment by cjw296** [comment]
Tangentially, @msuozzo: the "temporary" skip you introduced in https://github.com/python/cpython/pull/25326 is now almost a year old...

**Pull request merged**